### PR TITLE
External Source Maps!

### PR DESCRIPTION
**Pull request by dominicbarnes**
This adds support for external source-maps!

The `--development` flag no longer _implies_ source-maps, instead the `--source-maps` (or `--inline-source-maps`) flags must be used.

`Duo#sourceMap(value)` has been added to the API.
 * when `value === true`, an external source-map will be created. (ex: `build/index.js`
+ `build/index.js.map`)
 * when `value === 'inline'`, the output file will have the sourcemap added as
an inline comment.
 * when `value === false`, no source-maps will be generated at all.

In the CLI, when printing to `stdout`, only inline source-maps are supported. Thus, if you only specify `true`, it will automatically change to `inline`.

**BREAKING:** this changes the api for `Duo#run()`, it now returns the same object that `duo-pack@2` returns. (ie: `{ code, map }`)

### CODE REVIEW

**Comment by dominicbarnes** [comment]
I'm going to work on adding more tests for the CLI, so hold off on merging this quite yet. But feel free to discuss and suggest changes right now.

/cc @stephenmathieson @MatthewMueller @yields

**Comment by dominicbarnes** [comment]
Yay, the tests are failing for node 0.10 (only examples it looks like)

I'll have to look into this later tonight.

**Comment by stephenmathieson** [comment]
imo, `--development` should still imply source maps

File: bin/_duo [hunk 1]
```diff
@@ -66,6 +67,8 @@ var program = new Command('duo')
   .option('-C, --no-cache', 'enable or disable the cache during build', true)
   .option('-d, --development', 'include development dependencies.', false)
   .option('-g, --global <name>', 'expose entry as a global <name>', null)
+  .option('-i, --inline-source-map', 'include source-maps inline to the entry file', false)
```
Comment (stephenmathieson) [inline bin/_duo]: what about `--source-map={filepath,inline}`?

**Comment by stephenmathieson** [comment]
This is awesome @dominicbarnes! I left a few comments, but overall this lgtm!

**Comment by cristiandouce** [comment]
IMHO atomic options are better.

As the description already says: `--development` just implies `include development dependencies`

Then, if you want development dependencies and sourcemaps you can still:

`duo --development --source-maps entry.js`

I believe the more verbose the options are the better.

File: bin/_duo (hunk 1 above)
Comment (dominicbarnes) [inline bin/_duo]: I preferred to not have to specify a source-map path, having it automatically drop it in the right place was more important to me. (imo)

**Comment by dominicbarnes** [comment]
I'm with @cristiandouce on this one, I prefer the more explicit options. (was wanting to revisit duo-pack itself to change that API to reflect this)

File: bin/_duo (hunk 1 above)
Comment (stephenmathieson) [inline bin/_duo]: hmm okay, no big deal :)

I just thought two source map options may be a bit much :p

**Comment by stephenmathieson** [comment]
@dominicbarnes @cristiandouce is there ever a time when you'd want dev deps installed but *not* want source maps included?

**Comment by cristiandouce** [comment]
@stephenmathieson maybe for testing in a CI? dev deps are installed when running assert and stuff, right?

**Comment by stephenmathieson** [comment]
@cristiandouce would you actually actually build a file for testing in a ci differently than in your browser?

this seems silly to me:

```makefile
test-browser: build/test.js
  duo test --build $< browser

test-ci: build/ci.js
  duo test --build $< phantomjs

build/test.js: test.js index.js
  duo --development --source-maps index.js > $@

build/ci.js: test.js index.js
  duo --development > $@
```

**Comment by yields** [comment]
nice!

though `--development` should implicitly add source-maps i think, maybe `--source-map` should override the inline source-map if needed.

**Comment by dominicbarnes** [comment]
Sounds like we have a bit of a tie going on here, would you like to weigh in @MatthewMueller? (ie: should `--development` automatically imply `--source-map`?)

**Comment by dominicbarnes** [comment]
And just to make sure everyone understands the exact semantics of what I've proposed here:

```sh
# generates build/index.js + build/index.js.map
duo --source-map index.js

# generates build/index.js + inline source-map
duo --inline-source-map index.js

# generates build/index.js (no source-maps)
duo index.js
```

I look around at a couple other libraries to see how they did this:

 * [Uglify](https://github.com/mishoo/UglifyJS2) only supports external source-maps via the `--source-map` flag (plus several other source-map related flags)
 * [CoffeeScript](http://coffeescript.org/) only supports external-source-maps via the `--map` option. (no other config)
 * [Browserify](https://github.com/substack/node-browserify#usage) in _core_ only supports inline source-maps. (via the `--debug` flag, no other config) They defer to another library ([exorcist](https://github.com/thlorenz/exorcist)) to generate external source-maps.
 * [Babel](https://babeljs.io/docs/usage/cli/) supports inline source-maps (via `--source-maps-inline`) and external source-maps (via `--source-maps`)

Looks like I adopted the same methodology as babel. (w/o even realizing it!)

**Comment by yields** [comment]
i agree with most of those changes, but i can't understand why the need for `--inline-source-map`?

seems like it would be nicer from a user point-of-view to have inlined source maps whenever `--development` is given, but also have `--source-map` to override the inlined ones...

_i might be missing something_

**Comment by stephenmathieson** [comment]
@yields i agree fully

**Comment by dominicbarnes** [comment]
@yields the main reason for the extra CLI flag was because having an optional value for a flag is tricky. At first I wanted to do `--source-maps [inline]`, which meant I wanted either `duo --source-maps index.js` or `duo --source-maps inline index.js`. Unfortunately, commander would confuse `index.js` for the that optional bit.

So, @yields and @stephenmathieson would like something like this:

```sh
# generates build/index.js + build/index.js.map
duo --source-map index.js
duo --development --source-map index.js

# generates build/index.js + inline source-map
duo --development index.js

# generates build/index.js (no source-maps)
duo index.js
```

I think I still prefer the more explicit options, but I can live with either! :)

**Comment by dominicbarnes** [comment]
Oh btw, I've made a slight change. I've renamed the flags to be plural (since multiple entries can be processed at once)

**Comment by dominicbarnes** [comment]
I just had a compromise/idea, what if we support both `--development` _and_ `--inline-source-map`?

```sh
# generates build/index.js + build/index.js.map
duo --source-maps index.js
duo --development --source-maps index.js

# generates build/index.js + inline source-map
duo --development index.js
duo --inline-source-maps index.js

# generates build/index.js (no source-maps)
duo index.js
```

Basically, what if `--development` just acts as a "soft flag" for "inline source maps and development dependencies"?

**Comment by cristiandouce** [comment]
@stephenmathieson what I mean is that I'd avoide sourcemaps or inline sourcemaps when building `build/test.js`. Yet, nevermind, It's not a big issue to me.

I'd just prefer to avoide the "implies" which rely hidden on documentation which you only check when the command fails.

If the option `--development` implies sourcemaps, I'd good to add the implication at it's command line description, so when `duo --help` I see exactly what the development flag does.

This seems silly to me:

```
# only downloads dev deps, no source-maps
duo --development --no-inline-source-maps index.js
```

having to exclude some hidden default. It's more explicit to add flags that add functionality than remove. IMHO

**Comment by dominicbarnes** [comment]
Hey guys, it's been almost a week and I haven't heard anything new. Since we seem to be divided, would you guys be willing to merge this as-is, and we can change it later if it turns out to be a problem?

**Comment by dominicbarnes** [comment]
I'd really like to hear from @MatthewMueller regarding the discussions above, but I will probably proceed with merging this sometime this weekend if I don't hear anything else.

**Comment by MatthewMueller** [comment]
Didn't mean to leave this hanging @dominicbarnes! I'm tentatively with @yields in that `--development` should imply source-maps and we should make smart choices for the developer about the options of `--development`.

That being said, I'm not familiar with all the pros and cons of external vs internal source maps, but I think we should take a stand on how development "should" be. in other words, we should be weighing the pros and cons and making that decision so newcomers don't have to. I think what you suggested @dominicbarnes in your "compromise/idea" would work.

**Comment by dominicbarnes** [comment]
@duojs/owners I'm going to try and wrap this up today, here's what the final API should look like:

 ```sh
# generates build/index.js + build/index.js.map
duo --external-source-maps index.js
duo --development --external-source-maps index.js

# generates build/index.js + inline source-map
duo --development index.js

# generates build/index.js (no source-maps)
duo index.js
```

This preserves the current behavior of `--development` (that implies _inline_ source maps) while allowing for external source-maps with an extra flag. Success!

**Pull request merged**